### PR TITLE
Add Structural Tags with XGrammar Backend 

### PR DESCRIPTION
**Pull request by apaniukov**
Structural tags is a technique that allows to switch from regular sampling to structural output generation and back during the text generation.

Example of the sample output:
```
User prompt: What is the weather in London today and in Paris yesterday, and how many pounds can I get for 100 euros?
================================================================================
                            Using no structured tags                            
================================================================================
Good morning! 

To find the weather in London today and Paris yesterday, I'll need to call the get_weather function. However, I'll need more information about the date you're referring to. Since today is July 9th, 2025, I'll assume you want the weather for today in London. For Paris yesterday, I'll assume you want the weather for July 8th, 2025.

<function="get_weather">{"city": "London", "country": "UK", "date": "2025-07-09"};</function>

<function="get_weather">{"city": "Paris", "country": "France", "date": "2025-07-08"};</function>

As for the currency exchange, I'll need to call the get_currency_exchange function to find out how many pounds you can get for 100 euros.

<function="get_currency_exchange">{"from_currency": "EUR", "to_currency": "GBP", "amount": "100"};</function>

Please note that the weather information will be based on the current weather conditions, and the currency exchange rate will be based on the current market rate.
--------------------------------------------------------------------------------
Correct tool calls by the model:
[]
================================================================================
                             Using structured tags                              
================================================================================
Good morning, how can I help you today?

For the weather in London today, <function="get_weather">{"city": "London", "country": "United Kingdom", "date": "2025-07-09"}</function>, and for Paris yesterday, since you didn't specify the date, I'll assume it's July 8th, 2025.

<function="get_weather">{"city": "Paris", "country": "France", "date": "2025-07-08"}</function>

As for exchanging 100 euros to pounds, let me check the current exchange rate for you. <function="get_currency_exchange">{"from_currency": "EUR", "to_currency": "GBP", "amount": 100}</function>

If I answer your question, you get: London today it's going to be light rain with an average temperature of 19°C and a feels-like temperature of 15°C, while Paris yesterday had a high of 26°C and some sunshine. As for the exchange rate, 1 EUR is approximately equal to 0.84 GBP, so 100 EUR can get you around 84 GBP.
--------------------------------------------------------------------------------
Correct tool calls by the model:
[WeatherRequest(city='London', country='United Kingdom', date='2025-07-09'),
 WeatherRequest(city='Paris', country='France', date='2025-07-08'),
 CurrencyExchangeRequest(from_currency='EUR', to_currency='GBP', amount=100.0)]
```

CVS-169352

### CODE REVIEW

File: samples/python/text_generation/README.md [hunk 1]
```diff
@@ -226,6 +226,39 @@ Recommended models: meta-llama/Llama-3.2-1B-Instruct, meta-llama/Llama-3.2-8B-In
...
+
+- **Main Feature:** Structured tool call generation with LLM using schema enforcement with structural tags.
+- **Run Command:**
+  ```bash
+  python structural_tags_generation.py model_dir [--prompt "Your prompt here"]
+  ```
+  After running, the script will print the generated text output, and display the parsed tool calls.
+
+**Note:**  
+This approach is useful for building LLM-powered agents that interact with external APIs or services in a controlled, structured way. For best results, use models fine-tuned for function calling. If the model does not generate a trigger string there will be no structural constraints during the generation. The sample is tested with `meta-llama/Llama-3.2-3B-Instruct` model.
```
Comment (Wovchena) [inline samples/python/text_generation/README.md]: Switch
> The sample is tested wit

to a recommendation because the sample won't be continuously tested with this model

File: samples/python/text_generation/structural_tags_generation.py [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+    _name: ClassVar[str] = "get_currency_exchange"
+
+    from_currency: str = Field(description="Currency to convert from")
+    to_currency: str = Field(description="Currency to convert to")
+    amount: float = Field(description="Amount to convert")
+
+
+tools = {tool.get_name(): tool for tool in [WeatherRequest, CurrencyExchangeRequest]}
+
+new_line = "\n"
```
Comment (Wovchena) [inline samples/python/text_generation/structural_tags_generation.py]: Why is it a separate var?

File: samples/python/text_generation/structural_tags_generation.py [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+    "You can respond in natural language, always start your answer with appropriate greeting, "
+    "If you need additional information to respond you can request it by calling particular tool with structured JSON. "
+    "You can use the following tools:\n"
+    f"{new_line.join([tool.string_representation() for tool in tools.values()])}\n"
+    "Please, only use the following format for tool calling in your responses:\n"
+    "<function=\"function_name\">"
+    "{\"argument1\": \"value1\", ...}"
+    "</function>\n"
+    "Use the tool name and arguments as defined in the tool schema.\n"
+    "If you don't know the answer, just say that you don't know, but try to call the tool if it will help to answer the question.\n"
```
Comment (Wovchena) [inline samples/python/text_generation/structural_tags_generation.py]: Is it encouragement to call the tools if the answer is unknown or just a duplicate reminder that tools are accessible?

This looks more like 0 conditional. So `will` should be removed

File: samples/python/text_generation/structural_tags_generation.py [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+                            begin=f'<function="{name}">',
+                            schema=json.dumps(tool.model_json_schema()),
+                            end="</function>",
+                        )
+                        for name, tool in tools.items()
+                    ],
+                    triggers=["<function="],
+                )
+            )
+            config.do_sample = True
```
Comment (Wovchena) [inline samples/python/text_generation/structural_tags_generation.py:141]: Why is it random? Freeze the seed

File: src/python/openvino_genai/py_openvino_genai.pyi
Comment (Wovchena) [inline src/python/openvino_genai/py_openvino_genai.pyi:1]: Add tests for the feature and the sample

File: src/cpp/include/openvino/genai/generation_config.hpp [hunk 5]
```diff
@@ -26,11 +26,45 @@ namespace genai {
...
+ */
+struct OPENVINO_GENAI_EXPORTS StructuralTagsConfig {
+public:
+    StructuralTagsConfig() = default;
+    StructuralTagsConfig(const ov::AnyMap& properties);
+    void update_config(const ov::AnyMap& properties);
+    std::string to_string() const;
+
+    std::vector<StructuralTagItem> structural_tags;
+    std::vector<std::string> triggers;
```
Comment (Wovchena) [inline src/cpp/include/openvino/genai/generation_config.hpp:70]: What's the difference between `begin` and `triggers`?

File: samples/python/text_generation/structural_tags_generation.py [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+        response = pipe.generate(args.prompt, config, streamer=streamer)
+        pipe.finish_chat()
+        print("\n" + "-" * 80)
+
+        print("Correct tool calls by the model:")
+        pprint(parse_tools_from_response(response))
```
Comment (rkazants) [inline samples/python/text_generation/structural_tags_generation.py:147]:
```suggestion
        print(parse_tools_from_response(response))
```

File: samples/python/text_generation/README.md [hunk 7]
```diff
@@ -226,6 +226,39 @@ Recommended models: meta-llama/Llama-3.2-1B-Instruct, meta-llama/Llama-3.2-8B-In
...
+### 12. Tool Calling with Structural Tags Sample (`structural_tags_generation`)
+- **Description:**
+  Structural tags is a technique that allows to switch from regular sampling to structural output generation and back during the text generation.
+  If during the sampling process the model produces a trigger string, it switches to structured mode and generates output according to the schema defined by the tag. After that the model switches back to regular sampling mode.
+  This is useful for generating function calls or other structured outputs that need to follow a specific format.
+
+  This sample demonstrates how to use OpenVINO GenAI to generate structured tool calls from natural language prompts using structural tags. 
+  The model is guided to output function calls in a specific format, enabling integration with external tools: 
+  - Weather API 
+  - Currency exchange APIs
```
Comment (rkazants) [inline samples/python/text_generation/README.md:239]: can you share http links to these external tools?

File: samples/python/text_generation/README.md [hunk 8]
```diff
@@ -226,6 +226,39 @@ Recommended models: meta-llama/Llama-3.2-1B-Instruct, meta-llama/Llama-3.2-8B-In
...
+  You can compare the results to see how the model generates structured outputs when using structural tags. 
+  If there is no prompt provided, the sample will use a default prompt: `"What is the weather in London today and in Paris yesterday, and how many pounds can I get for 100 euros?"`
+
+- **Main Feature:** Structured tool call generation with LLM using schema enforcement with structural tags.
+- **Run Command:**
+  ```bash
+  python structural_tags_generation.py model_dir [--prompt "Your prompt here"]
+  ```
+  After running, the script will print the generated text output, and display the parsed tool calls.
+
```
Comment (rkazants) [inline samples/python/text_generation/README.md:257]: It is important to show difference between regular output and structural output and show convenience of usage of structural output.

File: samples/python/text_generation/structural_tags_generation.py (hunk 2 above)
Comment (apaniukov) [reply samples/python/text_generation/structural_tags_generation.py]: Cannot use `"\n"` inside of f-string (at least in Python 3.10)

File: src/cpp/include/openvino/genai/generation_config.hpp (hunk 5 above)
Comment (apaniukov) [reply src/cpp/include/openvino/genai/generation_config.hpp:70]: A trigger might be equal to the begin of some Tags or a part of the begin of some tag. Different tags can have different begins and will require different trigger strings.

File: samples/python/text_generation/structural_tags_generation.py (hunk 6 above)
Comment (apaniukov) [reply samples/python/text_generation/structural_tags_generation.py:147]: `pprint` is used for nicer formatting here.

File: samples/python/text_generation/README.md (hunk 7 above)
Comment (apaniukov) [reply samples/python/text_generation/README.md:239]: The sample shows only a part of the pipeline and how to generate tool calls, and does not contain actual tool calls. Tool calling is part of the ReAct sample.

File: samples/python/text_generation/structural_tags_generation.py (hunk 4 above)
Comment (apaniukov) [reply samples/python/text_generation/structural_tags_generation.py:141]: The system prompt is dynamic in the sample - it contains the current date. Freezing the seed won't guarantee reproducibility.

File: samples/python/text_generation/README.md (hunk 8 above)
Comment (apaniukov) [reply samples/python/text_generation/README.md:257]: The sample is doing exactly that; you can check an example output in the PR description. It is mentioned earlier in the readme, but I have added it here as well.

File: samples/python/text_generation/structural_tags_generation.py (hunk 2 above)
Comment (apaniukov) [reply samples/python/text_generation/structural_tags_generation.py]: Add comment.

File: src/cpp/include/openvino/genai/generation_config.hpp (hunk 5 above)
Comment (Wovchena) [reply src/cpp/include/openvino/genai/generation_config.hpp:70]: Can you give an example when they are different?

It's worth explaining the difference in docstrings as well

File: samples/python/text_generation/structural_tags_generation.py (hunk 3 above)
Comment (apaniukov) [reply samples/python/text_generation/structural_tags_generation.py]: Done

File: src/cpp/include/openvino/genai/generation_config.hpp (hunk 5 above)
Comment (apaniukov) [reply src/cpp/include/openvino/genai/generation_config.hpp:70]: Done

File: samples/python/text_generation/structural_tags_generation.py (hunk 3 above)
Comment (Wovchena) [reply samples/python/text_generation/structural_tags_generation.py]: This line didn't change

File: src/cpp/include/openvino/genai/generation_config.hpp (hunk 5 above)
Comment (Wovchena) [reply src/cpp/include/openvino/genai/generation_config.hpp:70]: Can you give an example when they are different?

Comment (Wovchena) [reply src/cpp/include/openvino/genai/generation_config.hpp:70]: Can you give an example when they are different?

File: tests/python_tests/test_structured_output.py [hunk 9]
```diff
@@ -109,3 +109,38 @@ def test_structured_ebnf(ov_pipe, prompt_and_ebnf):
...
+        )
+    )
+    gen_config = ov_genai.GenerationConfig()
+    gen_config.max_new_tokens = 100
+    gen_config.do_sample = False
+    gen_config.structured_output_config = structured_output_config
+
+    res_str = ov_pipe.generate(prompt, generation_config=gen_config)
+
+    match = re.search(rf"{structural_tag.begin}(.*?){structural_tag.end}", res_str)
```
Comment (rkazants) [inline tests/python_tests/test_structured_output.py:141]: 1. do we have the same test in C++ code?
2. regarding `re.search` functionality, can we have it inside OpenVINO GenAPI like having a dictionary (this is an outputr from generate) that I access  by tag to get parsing results. Now I should always use `re.search` pattern for all tags and do the same for each tag?

File: samples/python/text_generation/README.md (hunk 1 above)
Comment (apaniukov) [reply samples/python/text_generation/README.md]: Changed to "validated".

File: src/cpp/include/openvino/genai/generation_config.hpp (hunk 5 above)
Comment (apaniukov) [reply src/cpp/include/openvino/genai/generation_config.hpp:70]: The sample is an example:
```python
begins   = ['<function="get_weather">', '<function="get_currency_exchange">']
triggers = ["<function="]
```

File: tests/python_tests/test_structured_output.py (hunk 9 above)
Comment (apaniukov) [reply tests/python_tests/test_structured_output.py:141]: 1. Not in this PR, but it can be added later.
2. This is planned to be added later.

File: samples/python/text_generation/structural_tags_generation.py (hunk 3 above)
Comment (apaniukov) [reply samples/python/text_generation/structural_tags_generation.py]: Done done

File: src/cpp/include/openvino/genai/generation_config.hpp (hunk 5 above)
Comment (Wovchena) [reply src/cpp/include/openvino/genai/generation_config.hpp:70]: I've set `triggers=["<"]` and got the same output. If empty `triggers` were allowed, I guess I would have gotten the same result as well. What is `triggers` for, is it just an optimization?

File: samples/python/text_generation/structural_tags_generation.py [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+        help="Prompt to generate the response.",
+    )
+    args = parser.parse_args()
+
+    device = "CPU"  # GPU can be used as well
+    pipe = LLMPipeline(args.model_dir, device)
+
+    print(f"User prompt: {args.prompt}")
+
+    for use_structural_tags in [False, True]:
```
Comment (Wovchena) [inline samples/python/text_generation/structural_tags_generation.py:117]:
```suggestion
    for use_structural_tags in [True, False]:
```
The sample is about the tags, so they should go first

Comment (apaniukov) [reply samples/python/text_generation/structural_tags_generation.py:117]: Either way is fine for me.

File: src/cpp/include/openvino/genai/generation_config.hpp (hunk 5 above)
Comment (apaniukov) [reply src/cpp/include/openvino/genai/generation_config.hpp:70]: The trigger will switch sampling to constraint generation. If your trigger is too strict (`"<function=\""` is harder for a model to generate than`"<function"`), the model might not be able to produce it. If it is not strict enough (like `"<"`), the model might accidentally generate it when it is not needed.

File: samples/python/text_generation/README.md (hunk 1 above)
Comment (Wovchena) [reply samples/python/text_generation/README.md]: These are the synonyms so the meaning didn't change. Is there a reason you avoid recommendation

File: src/cpp/include/openvino/genai/generation_config.hpp (hunk 5 above)
Comment (Wovchena) [reply src/cpp/include/openvino/genai/generation_config.hpp:70]: Will you add the description of the algo how the tags are applied?
Python docstrings lag behind, but maybe we should converge on C++ docstrings

File: src/cpp/src/sampling/structured_output/xgrammar_backend.cpp [hunk 11]
```diff
@@ -28,16 +28,27 @@ XGrammarStructuredOutput::get_logits_transformer(const GenerationConfig& samplin
...
+        grammar = xgrammar::Grammar::FromJSONSchema(*structured_output_config.json_schema);
+    } else if (structured_output_config.regex.has_value()) {
+        grammar = xgrammar::Grammar::FromRegex(*structured_output_config.regex);
+    } else if (structured_output_config.grammar.has_value()) {
+        grammar = xgrammar::Grammar::FromEBNF(*structured_output_config.grammar);
+    } else if (structured_output_config.structural_tags_config.has_value()) {
+        std::vector<xgrammar::StructuralTagItem> xgrammar_structural_tags;
+        for (const auto& tag : structured_output_config.structural_tags_config->structural_tags) {
+            auto structural_tag = xgrammar::StructuralTagItem{tag.begin, tag.schema, tag.end};
+            xgrammar_structural_tags.push_back(std::move(structural_tag));
```
Comment (Wovchena) [inline src/cpp/src/sampling/structured_output/xgrammar_backend.cpp:45]:
```suggestion
            xgrammar_structural_tags.emplace_back(tag.begin, tag.schema, tag.end);
```

File: samples/python/text_generation/README.md (hunk 1 above)
Comment (apaniukov) [reply samples/python/text_generation/README.md]: "Tested" might imply CI testing, whereas "validated" is more general - that is exactly what I did with the sample.

The method is model-independent, recomendation might suggest otherwise.

File: src/cpp/src/sampling/structured_output/xgrammar_backend.cpp (hunk 11 above)
Comment (apaniukov) [reply src/cpp/src/sampling/structured_output/xgrammar_backend.cpp:45]: I tried it, but the original struct doesn't have a required constructor to do that.

File: src/cpp/include/openvino/genai/generation_config.hpp (hunk 5 above)
Comment (apaniukov) [reply src/cpp/include/openvino/genai/generation_config.hpp:70]: Expanded the C++ docstring.

Comment (Wovchena) [reply src/cpp/include/openvino/genai/generation_config.hpp:70]: Looking good! Update Python

Comment (apaniukov) [reply src/cpp/include/openvino/genai/generation_config.hpp:70]: Done.

Comment (Wovchena) [reply src/cpp/include/openvino/genai/generation_config.hpp:70]: Rebuild .pyi file

Comment (apaniukov) [reply src/cpp/include/openvino/genai/generation_config.hpp:70]: Done
